spec/topological_inventory/orchestrator/metric_scaler/prometheus_watcher_spec.rb

```diff
@@ -0,0 +1,123 @@
+require "topological_inventory/orchestrator/metric_scaler/prometheus_watcher"
+require "topological_inventory/orchestrator/test_models/object_manager"
+
+describe TopologicalInventory::Orchestrator::MetricScaler::PrometheusWatcher do
+  let(:logger) { Logger.new(StringIO.new).tap { |logger| allow(logger).to receive(:info) } }
+
+  let(:annotations) do
+    {
+      "metric_scaler_max_replicas"        => "10",
+      "metric_scaler_min_replicas"        => "1",
+      "metric_scaler_target_usage_pct"    => "1",
+      "metric_scaler_scale_threshold_pct" => ".5",
+    }
+  end
+
+  let(:deployment_config_name) { 'topological-inventory-persister' }
+  let(:deployment)     { double("deployment", :metadata => double("metadata", :name => deployment_config_name, :annotations => annotations), :spec => double("spec", :replicas => replicas)) }
+  let(:metrics)        { watcher.send(:metrics) }
+  let(:object_manager) { double("TopologicalInventory::Orchestrator::ObjectManager", :get_deployment_configs => [deployment], :get_deployment_config => deployment) }
+  let(:replicas)       { 3 }
+  let(:watcher)        { described_class.new(deployment, deployment.metadata.name, 'prometheus.mnm-ci', logger) }
+
+  before { allow(TopologicalInventory::Orchestrator::ObjectManager).to receive(:new).and_return(object_manager) }
+
+  context "downloading metrics" do
+    before do
+      stub_const("#{described_class}::METRICS_CHECK_INTERVAL", 0)
+
+      # Allow just a single run
+      allow(watcher).to receive(:finished?) do
+        finished = watcher.instance_variable_get(:@finished)
+        old_value = finished.value
+        finished.value = true
+        old_value
+      end
+    end
+
+    it "fills metrics array when successful" do
+      downloaded = 10.times.collect { |i| [Time.now.to_i, 10]}
+      allow(watcher).to receive(:download_metrics).and_return(downloaded)
+
+      watcher.start
+      watcher.thread.join
+
+      expect(metrics.values).to eq(Array.new(10, 10))
+    end
+
+    it "allows scaling run only when successful" do
+      downloaded = 5.times.collect { |i| [Time.now.to_i, 10]}
+      allow(watcher).to receive(:download_metrics).and_return(downloaded)
+
+      watcher.start
+      watcher.thread.join
+
+      expect(metrics.average).to eq(10)
+      expect(watcher.scaling_allowed?).to be_truthy
+    end
+
+    it "doesn't allow scaling run when no data collected" do
+      allow(watcher).to receive(:download_metrics).and_return([])
+
+      watcher.start
+      watcher.thread.join
+
+      expect(watcher.scaling_allowed?).to be_falsey
+    end
+
+    it "doesn't allow scaling run when zero data collected" do
```

Comment:
This is a great spec - which brings up my question above. What if we can't get metrics from Prometheus due to a rescued exception? Should there be an alert for that?  

If there wasn't an exception then there seems to be potential for a lot of lag build up - and we wouldn't know that the Orchestrator couldn't see that and act accordingly.